rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1182 @@
+  - The cutoff is usually determined empirically based on common usage patterns.
+
+**Examples:**
+
+```nix
+{
+  #1
+  buildInputs = [
+    foo
+    bar
+    baz
+
+    somethingElse
+  ];
+
+  #2
+  systemd.services = {
+    foo = { };
+    bar = { };
+  };
+
+  #3
+  inherit
+    lib
+    foo
+    bar
+    baz
+    ;
+}
+```
+
+### Function application
+
+**Description:**
+
+- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
+- The last argument receives special treatment, to better represent common coding patterns.
+- As many arguments as possible are fit onto the first line.
+  - If all but the last argument do fit, then the last argument may start on the same line.
+  - If an earlier argument does not fit onto the first line, then itself and all the following ones start on a new line. This is called the expanded form.
+  - All arguments that are not on the same line as the function are indented by one level.
+
+**Examples:**
+
+```nix
+#1
+function arg1 arg2
+
+#2
+function arg1 arg2 {
+  more = "things";
+}
+
+#3
+function arg1 arg2 arg3 # reached line limit here
+  arg4
+  arg5
+  [
+    1
+    2
+  ]
+  arg7
+
+#4 All arguments that don't fit on the same line, start indented on a new line
+function arg1
+  {
+    # items that don't fit on same line
+  }
+  arg3
+
+#5
+function
+  {
+    a = 1;
+    b = 2;
+  }
+  {
+    c = 1;
+    d = 2;
+  }
+
+#6 (with "--width=20")
+function arg1 (
+  function2 args
+)
+```
+
+**Drawbacks**
+
+- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.
+
+**Rationale and alternatives**
+
+- Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.
+- Compacting multiline arguments like this:
+  ```nix
+  #4b
+  function arg1 {
+    # stuff
+  } arg3
+
+  #5b
+  function {
+    # …
+  } {
+    # …
+  }
+  ```
+  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
+  - This does not work well with line length limits on short arguments like in example #3.
+
+### Function declaration
+
+**Description**
+
+- The body of the function is not indented relative to its arguments.
+- Multiple ("simple") identifier arguments are written onto the same line if possible.
+- Attribute set arguments always start on a new line; they are not mixed with identifier arguments.
+  - If they have few attributes, the argument may be written on a single line, otherwise the expanded form is used.
+- Attribute set arguments have their attributes on a new line each with indentation, followed by a trailing comma.
+
+**Examples**
+
+```nix
+#1
+name: value: name ++ value
+
+#2
+name: value:
+{
+  "${name}-foo" = value;
+}
+
+#3
+{ pkgs }: pkgs.hello
+
+#4
+args@{
```

Comment:
Does `args@` always come at the front? I'm asking because I usually do `{ whatever }@args` which is also valid syntax.

Reply:
Because both is allowed in the Nix syntax, the formatter must also support both. I guess this should be a general rule written down somewhere.